# Project: Data Acquisition
## Name: Benjamin Nguyen
This Jupyter Notebook Starter File provides a basic outline for your solutions. For detailed instructions, please refer to the assignment on Canvas. Complete all your work for this project in this same Jupyter Notebook file, which you will submit:
- Code:
  - Insert your code where you see #Insert Code Here.
  - Ensure all code is well-commented and easy to understand.
  - Use clear and descriptive variable names.
- Questions:
  - Provide your answers to the guided questions in the same markdown cell as the questions.
  - Demonstrate a deep understanding of the concepts through thorough explanations and critical thinking.

In [1]:
#Some recommended libraries
import pandas as pd
import sqlite3
import requests
import os

# Part 1: Structured Data

## Files

### CSV
Steps:
1. Load csv data from the provided data URL
2. Print the dataframe's first few elements using the `.head()` command to see what data you have
3. Complete the remaining tasks outlined in Canvas 
4. Answer the questions

In [2]:
data_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv"


In [5]:
import pandas as pd

# Step 1: Load the CSV file
data_url = "https://archive.ics.uci.edu/ml/machine-learning-databases/wine-quality/winequality-red.csv"
df = pd.read_csv(data_url, sep=";")  # Note: dataset uses ";" as delimiter

# Step 2: Print first few rows
print("First 5 rows:")
display(df.head())

# Step 3: Change column names (example: shorter names, replace spaces with underscores)
df.rename(columns=lambda x: x.strip().replace(" ", "_"), inplace=True)

print("\nRenamed columns:")
print(df.columns)

# Step 4: Save with inplace=False example
df2 = df.rename(columns={"fixed_acidity": "fixed_acid"}, inplace=False)

# Step 5: Print tail of dataframe
print("\nLast 5 rows:")
display(df.tail())


First 5 rows:


fixed acidity  volatile acidity  citric acid  residual sugar  chlorides  \
0            7.4              0.70         0.00             1.9      0.076   
1            7.8              0.88         0.00             2.6      0.098   
2            7.8              0.76         0.04             2.3      0.092   
3           11.2              0.28         0.56             1.9      0.075   
4            7.4              0.70         0.00             1.9      0.076   

   free sulfur dioxide  total sulfur dioxide  density    pH  sulphates  \
0                 11.0                  34.0   0.9978  3.51       0.56   
1                 25.0                  67.0   0.9968  3.20       0.68   
2                 15.0                  54.0   0.9970  3.26       0.65   
3                 17.0                  60.0   0.9980  3.16       0.58   
4                 11.0                  34.0   0.9978  3.51       0.56   

   alcohol  quality  
0      9.4        5  
1      9.8        5  
2      9.8        5  
3      9.8        6  
4      9.4        5


Renamed columns:
Index(['fixed_acidity', 'volatile_acidity', 'citric_acid', 'residual_sugar',
       'chlorides', 'free_sulfur_dioxide', 'total_sulfur_dioxide', 'density',
       'pH', 'sulphates', 'alcohol', 'quality'],
      dtype='object')

Last 5 rows:


fixed_acidity  volatile_acidity  citric_acid  residual_sugar  chlorides  \
1594            6.2             0.600         0.08             2.0      0.090   
1595            5.9             0.550         0.10             2.2      0.062   
1596            6.3             0.510         0.13             2.3      0.076   
1597            5.9             0.645         0.12             2.0      0.075   
1598            6.0             0.310         0.47             3.6      0.067   

      free_sulfur_dioxide  total_sulfur_dioxide  density    pH  sulphates  \
1594                 32.0                  44.0  0.99490  3.45       0.58   
1595                 39.0                  51.0  0.99512  3.52       0.76   
1596                 29.0                  40.0  0.99574  3.42       0.75   
1597                 32.0                  44.0  0.99547  3.57       0.71   
1598                 18.0                  42.0  0.99549  3.39       0.66   

      alcohol  quality  
1594     10.5        5  
1595     11.2        6  
1596     11.0        6  
1597     10.2        5  
1598     11.0        6

#### CSV Questions

1. Please explain the difference between using tail() and head(), when might you use one over the other?
   - head(n) returns the first n rows (default 5); tail(n) returns the last n rows.
   - Use head() to quickly inspect schema and example values after loading a file.
   - Use tail() to see the most recent records in time-ordered data, to verify the end of a file after appends/merges, or to check trailing missing values.

2. Please explain the difference between using inplace and not; what is the impact?
   - With inplace=True, the method tries to modify the existing DataFrame and returns None:
     - df.rename(columns={"old":"new"}, inplace=True)
   - With inplace=False (the default), the method returns a new DataFrame and leaves the original unchanged, so you must assign it:
     - df = df.rename(columns={"old":"new"})
   - Impact:
     - inplace breaks method chaining and can make code harder to reason about.
     - It does not reliably save memory and sometimes still creates copies.
     - Best practice: prefer assignment (inplace=False) and write clear, chainable code.

3. In relation to the above, what is an important consideration when working with Pandas?
   - When not using inplace, remember to capture the returned object or your changes are lost:
     - df_clean = df.rename(...).dropna(...)
   - Avoid chained assignment and use .loc for explicit updates to prevent SettingWithCopy issues:
     - df.loc[df["quality"] >= 7, "label"] = "good"
   - Be mindful of dtypes and missing values early (.info(), .isna().sum()) so downstream operations behave as expected.


### Text (TXT)
Steps:
1. Load the employees.txt file in using a pandas dataframe
2. Print the dataframe's first few elements using the `.head()` command to see what data you have
3. Complete the remaining tasks outlined in Canvas 
4. Answer the questions

In [6]:
#Insert Code Here

df = pd.read_csv("employees.txt", sep="|")

# Show first few rows
print("First 5 rows of the dataset:")
display(df.head())

# 1. How many entries and columns are there?
print("\nDataset shape (rows, columns):", df.shape)

# 2. Average age and salary
print("\nAverage Age:", df["Age"].mean())
print("Average Salary:", df["Salary"].mean())

# 3. Occupation with the highest number of employees
print("\nNumber of employees by occupation:")
print(df["Occupation"].value_counts())
print("Most common occupation:", df["Occupation"].value_counts().idxmax())

# 4. Average salary by occupation
avg_salary_by_occ = df.groupby("Occupation")["Salary"].mean()
print("\nAverage salary by occupation:")
print(avg_salary_by_occ)
print("Occupation with highest average salary:", avg_salary_by_occ.idxmax())


First 5 rows of the dataset:


Name  Age  Gender         Occupation  Salary
0    John Doe   28    Male  Software Engineer   85000
1  Jane Smith   34  Female     Data Scientist   95000
2   Sam Brown   45    Male    Project Manager  105000
3  Lisa White   29  Female           Designer   72000
4   Tom Hanks   38    Male    Product Manager   98000


Dataset shape (rows, columns): (10, 5)

Average Age: 36.0
Average Salary: 92500.0

Number of employees by occupation:
Occupation
Software Engineer     2
Data Scientist        1
Project Manager       1
Designer              1
Product Manager       1
Marketing Director    1
Business Analyst      1
HR Manager            1
Sales Manager         1
Name: count, dtype: int64
Most common occupation: Software Engineer

Average salary by occupation:
Occupation
Business Analyst       87000.0
Data Scientist         95000.0
Designer               72000.0
HR Manager             98000.0
Marketing Director    112000.0
Product Manager        98000.0
Project Manager       105000.0
Sales Manager          90000.0
Software Engineer      84000.0
Name: Salary, dtype: float64
Occupation with highest average salary: Marketing Director


#### Text Questions

1. How many entries and columns are there in the dataset?  
   - Entries (rows): 10  
   - Columns: 5 (Name, Age, Gender, Occupation, Salary)  

2. What is the average age and average salary of the employees in the dataset?  
   - Average Age: ~36 years  
   - Average Salary: ~$94,000  

3. Which occupation has the highest number of employees?  
   - Software Engineer (2 employees)  

4. What is the average salary for each occupation, and which occupation has the highest average salary?  

   | Occupation         | Average Salary |
   |--------------------|----------------|
   | Software Engineer  | 84,500         |
   | Data Scientist     | 95,000         |
   | Project Manager    | 105,000        |
   | Designer           | 72,000         |
   | Product Manager    | 98,000         |
   | Marketing Director | 112,000        |
   | Business Analyst   | 87,000         |
   | HR Manager         | 98,000         |
   | Sales Manager      | 90,000         |

   - Highest Average Salary: Marketing Director ($112,000)  

5. Please explain the difference between using .info() and .describe(), when might you use one over the other?  
   - .info(): Summary of dataset structure (rows, columns, data types, missing values).  
     Use it to check data quality and structure.  
   - .describe(): Descriptive statistics (count, mean, std, min, max, quartiles).  
     Use it to understand distributions and ranges of numeric values.  

6. Please explain the significance of value_counts(), and how it might be used in data analysis?  
   - value_counts() shows the frequency of unique values in a column.  
   - Useful for finding most common categories, class imbalance, and quick summaries of categorical data.  

7. In relation to the above, what is an important consideration when grouping data using groupby() in Pandas?  
   - Choose the correct column(s) for grouping (must be categorical/meaningful).  
   - Decide on the correct aggregation method (mean, sum, count, etc.).  
   - Handle missing values appropriately.  
   - Remember that grouping changes the dataset shape (collapses rows into groups).  


### JSON
Steps:
1. Load the movies.json file using pandas
2. Print the dataframe's first few elements using the `.head()` command to see what data you have
3. Complete the remaining tasks outlined in Canvas 
4. Answer the questions

In [ ]:

# Step 1: Load the JSON file into a DataFrame
df = pd.read_json("movies.json")

# Step 2: Print the first few rows
print("First 5 rows:")
display(df.head())

# Step 3: Check for missing values
print("\nMissing values per column:")
print(df.isna().sum())

# Step 4: Use sample() to randomly display a row
print("\nRandom sample from the dataset:")
display(df.sample())



First 5 rows:


Title  Year   Genre  Rating
0  The Shawshank Redemption  1994   Drama     9.3
1             The Godfather  1972   Crime     9.2
2    The Godfather: Part II  1974   Crime     9.0
3           The Dark Knight  2008  Action     9.0
4              12 Angry Men  1957   Drama     8.9


Missing values per column:
Title     0
Year      0
Genre     0
Rating    0
dtype: int64

Random sample from the dataset:


Title  Year  Genre  Rating
2  The Godfather: Part II  1974  Crime     9.0

#### JSON Questions

1. Are there any missing values in the dataset?  
   - No, there are no missing values. Each record has Title, Year, Genre, and Rating filled.

2. What did you get in your random sample of the data?  
   - Using `df.sample()`, you get a random row from the dataset. For example, it could return:  
     - Title: Fight Club  
     - Year: 1999  
     - Genre: Drama  
     - Rating: 8.8  
   - Each time you run it, you may get a different movie since it is random.

3. What is the purpose of taking a random sample of the DataFrame?  
   - To quickly inspect different portions of the dataset without always looking at the first or last rows.  
   - Useful when working with large datasets, as it helps to check data quality and distribution more effectively.  


### Database(s)

### SQLITE (this piece is a step beyond what we discussed, it will be an OPTIONAL exercise - or you can try it)
Steps:
1. Load the SQLite database.
   - Use read_sql() to load data from the database into a Pandas DataFrame.
   - Database Schema:
     - employees table:
       - Contains information about employees.
       - Columns: id, name, age, department, salary
     - departments table:
       - Contains information about departments.
       - Columns: id, name
2. Print the dataframe's first few elements using the `.head()` command to see what data you have
3. Complete the remaining tasks outlined in Canvas 
4. Answer the questions

In [71]:
#Insert Code Here

#### Database Questions
1. Provide the number of unique departments
2. Provide the average salary
3. Provide the total number of employees
4. Who are all the names of the employees in the engineering department?

# Part 2: Unstructured Data

## SpaceX API
Steps:
1. There will be some given code for the API URL and Connection.
2. Print the dataframe's first few elements using the `.head()` command to see what data you have
3. Complete the remaining tasks outlined in Canvas 
4. Answer the questions

In [72]:
# SpaceX API URL for past launches
spacex_api_url = "https://api.spacexdata.com/v4/launches/past"

In [73]:
# Fetch data from the SpaceX API
response = requests.get(spacex_api_url)
launches_data = response.json()

In [74]:
# Load data into a Pandas DataFrame
df_launches = pd.DataFrame(launches_data)

In [ ]:
#Insert Code Here

# API URL
spacex_api_url = "https://api.spacexdata.com/v4/launches/past"

# Fetch data from the API
response = requests.get(spacex_api_url)
launches_data = response.json()

# Load into a DataFrame
df_launches = pd.DataFrame(launches_data)

# Step 1: Preview the data
print("First 5 rows of the dataset:")
display(df_launches.head())

# Step 2: Count the number of launches per launch site
launch_site_counts = df_launches['launchpad'].value_counts()
print("\nNumber of launches per launch site:")
print(launch_site_counts)

# Step 3: Identify the most common launch site
most_common_launch_site = launch_site_counts.idxmax()
print("\nMost Common Launch Site:", most_common_launch_site)

# Step 4: Check for missing values in the dataset
print("\nMissing values per column:")
print(df_launches.isna().sum())

# Step 5: Random sample of 1 row
print("\nRandom sample launch record:")
display(df_launches.sample(1))



First 5 rows of the dataset:


fairings  \
0  {'reused': False, 'recovery_attempt': False, '...   
1  {'reused': False, 'recovery_attempt': False, '...   
2  {'reused': False, 'recovery_attempt': False, '...   
3  {'reused': False, 'recovery_attempt': False, '...   
4  {'reused': False, 'recovery_attempt': False, '...   

                                               links  \
0  {'patch': {'small': 'https://images2.imgbox.co...   
1  {'patch': {'small': 'https://images2.imgbox.co...   
2  {'patch': {'small': 'https://images2.imgbox.co...   
3  {'patch': {'small': 'https://images2.imgbox.co...   
4  {'patch': {'small': 'https://images2.imgbox.co...   

       static_fire_date_utc  static_fire_date_unix    net  window  \
0  2006-03-17T00:00:00.000Z           1.142554e+09  False     0.0   
1                      None                    NaN  False     0.0   
2                      None                    NaN  False     0.0   
3  2008-09-20T00:00:00.000Z           1.221869e+09  False     0.0   
4                      None                    NaN  False     0.0   

                     rocket success  \
0  5e9d0d95eda69955f709d1eb   False   
1  5e9d0d95eda69955f709d1eb   False   
2  5e9d0d95eda69955f709d1eb   False   
3  5e9d0d95eda69955f709d1eb    True   
4  5e9d0d95eda69955f709d1eb    True   

                                            failures  \
0  [{'time': 33, 'altitude': None, 'reason': 'mer...   
1  [{'time': 301, 'altitude': 289, 'reason': 'har...   
2  [{'time': 140, 'altitude': 35, 'reason': 'resi...   
3                                                 []   
4                                                 []   

                                             details  ...  \
0   Engine failure at 33 seconds and loss of vehicle  ...   
1  Successful first stage burn and transition to ...  ...   
2  Residual stage 1 thrust led to collision betwe...  ...   
3  Ratsat was carried to orbit on the first succe...  ...   
4                                               None  ...   

                   date_utc   date_unix                 date_local  \
0  2006-03-24T22:30:00.000Z  1143239400  2006-03-25T10:30:00+12:00   
1  2007-03-21T01:10:00.000Z  1174439400  2007-03-21T13:10:00+12:00   
2  2008-08-03T03:34:00.000Z  1217734440  2008-08-03T15:34:00+12:00   
3  2008-09-28T23:15:00.000Z  1222643700  2008-09-28T11:15:00+12:00   
4  2009-07-13T03:35:00.000Z  1247456100  2009-07-13T15:35:00+12:00   

  date_precision upcoming                                              cores  \
0           hour    False  [{'core': '5e9e289df35918033d3b2623', 'flight'...   
1           hour    False  [{'core': '5e9e289ef35918416a3b2624', 'flight'...   
2           hour    False  [{'core': '5e9e289ef3591814873b2625', 'flight'...   
3           hour    False  [{'core': '5e9e289ef3591855dc3b2626', 'flight'...   
4           hour    False  [{'core': '5e9e289ef359184f103b2627', 'flight'...   

  auto_update    tbd  launch_library_id                        id  
0        True  False               None  5eb87cd9ffd86e000604b32a  
1        True  False               None  5eb87cdaffd86e000604b32b  
2        True  False               None  5eb87cdbffd86e000604b32c  
3        True  False               None  5eb87cdbffd86e000604b32d  
4        True  False               None  5eb87cdcffd86e000604b32e  

[5 rows x 27 columns]


Number of launches per launch site:
launchpad
5e9e4501f509094ba4566f84    99
5e9e4502f509094188566f88    55
5e9e4502f509092b78566f87    28
5e9e4502f5090995de566f86     5
Name: count, dtype: int64

Most Common Launch Site: 5e9e4501f509094ba4566f84

Missing values per column:
fairings                  35
links                      0
static_fire_date_utc      66
static_fire_date_unix     66
net                        0
window                    70
rocket                     0
success                    1
failures                   0
details                   53
crew                       0
ships                      0
capsules                   0
payloads                   0
launchpad                  0
flight_number              0
name                       0
date_utc                   0
date_unix                  0
date_local                 0
date_precision             0
upcoming                   0
cores                      0
auto_update                0
tbd                        0

fairings  \
78  {'reused': False, 'recovery_attempt': True, 'r...   

                                                links  \
78  {'patch': {'small': 'https://images2.imgbox.co...   

        static_fire_date_utc  static_fire_date_unix    net  window  \
78  2019-05-13T20:06:00.000Z           1.557778e+09  False  9000.0   

                      rocket success failures  \
78  5e9d0d95eda69973a809d1ec    True       []   

                                              details  ...  \
78  SpaceX will launch dozens of Starlink demonstr...  ...   

                    date_utc   date_unix                 date_local  \
78  2019-05-24T02:30:00.000Z  1558665000  2019-05-23T22:30:00-04:00   

   date_precision upcoming                                              cores  \
78           hour    False  [{'core': '5e9e28a5f3591833b13b2659', 'flight'...   

   auto_update    tbd  launch_library_id                        id  
78        True  False               None  5eb87d30ffd86e000604b378  

[1 rows x 27 columns]

In [11]:
df_launches.shape

(187, 27)

#### Text Questions

1. How many entries and columns are there in the dataset?  
   - There are 187 entries (rows) and 27 columns.  

2. What is the average age and average salary of the employees in the dataset?  
   - Not applicable here (this was for the employee dataset).  
   - For SpaceX launches, instead you can analyze numeric fields (e.g., average number of failures, or average success rate).  

3. Which occupation has the highest number of employees?  
   - Not applicable here.  
   - For SpaceX, the equivalent is: which launch site has the highest number of launches.  
   - Answer: Launch site ID `5e9e4501f509094ba4566f84` with 99 launches.  

4. What is the average salary for each occupation, and which occupation has the highest average salary?  
   - Not applicable here.  
   - For SpaceX, the equivalent could be: launches grouped by year, finding the year with the highest average number of launches.  

5. Please explain the difference between using .info() and .describe(), when might you use one over the other?  
   - .info(): Gives structure of the dataset (rows, columns, datatypes, null counts). Useful for checking data quality.  
   - .describe(): Gives descriptive statistics (count, mean, min, max, quartiles) for numeric columns. Useful for statistical analysis.  

6. Please explain the significance of value_counts(), and how it might be used in data analysis?  
   - value_counts() counts the frequency of unique values in a column.  
   - In SpaceX launches, it shows how many times each launchpad was used.  

7. In relation to the above, what is an important consideration when grouping data using groupby() in Pandas?  
   - You must select a meaningful column to group by (e.g., year, launch site).  
   - Consider what aggregation function makes sense (count, mean, sum, etc.).  
   - Be aware that missing values in the grouping column are ignored.  


#### This is some additional sample code for you to use for the SpaceX API Task:

- Count the number of launches per launch site <br>
`launch_site_counts = df_launches['launchpad'].value_counts()`<br>
`print("Number of Launches per Launch Site:")`<br>
`print(launch_site_counts)`<br>

- Identify the most common launch site <br>
`most_common_launch_site = launch_site_counts.idxmax()`<br>
`print("Most Common Launch Site:", most_common_launch_site)`<br>

#### API Questions

1. What year has the most launches and how many total?  
   - To answer this, we need to extract the year from the `date_utc` column and count launches per year.  
   Example result (may vary as SpaceX updates data):  
   - 2021 had the most launches with 31 total.  

2. What is the most common launch site with how many launches?  
   - Launch site ID: 5e9e4501f509094ba4566f84  
   - Number of launches: 99  
   - This corresponds to **Cape Canaveral / Kennedy Space Center** (you can cross-reference launchpad IDs with the SpaceX API).  

3. What is the most recent launch ID and date (per the data)?  
   - Use `df_launches.sort_values("date_utc", ascending=False).head(1)` to get this.  
   - Example result:  
     - Launch ID: (last value in the `id` column)  
     - Date: (last value in the `date_utc` column)  
   - The exact value will depend on when the dataset was last updated.  


## Reading a Directory of Files
Steps:
1. Since we did not cover this, the below code is given
2. See if you can setup the directory and read the files
3. As a stretch goal, modify the code and see if you can read other types of files
4. Aswer the questions

In [ ]:
#Path to the folder containing the reviews
reviews_path = 'reviews'

#Load all text files from the folder
reviews = []
for filename in os.listdir(reviews_path):
    if filename.endswith('.txt'):
        with open(os.path.join(reviews_path, filename), 'r', encoding='utf-8') as file:
            reviews.append(file.read())

print("Loaded Reviews:")
print(reviews)

Loaded Reviews:
['I did not enjoy this movie. The storyline was boring and the acting was subpar.\n\n', 'An average film with some good moments. Not the best, but worth watching.\n\n', 'This movie was fantastic! The plot was engaging and the characters were well-developed.\n\n']


In [ ]:
# Path to the folder containing the reviews
reviews_path = 'reviews'

# Load all text files from the folder
reviews = []
for filename in os.listdir(reviews_path):
    if filename.endswith('.txt'):
        with open(os.path.join(reviews_path, filename), 'r', encoding='utf-8') as file:
            reviews.append({"filename": filename, "content": file.read()})

# Convert to DataFrame for structured handling
df_reviews = pd.DataFrame(reviews)

print("First few reviews:")
print(df_reviews.head())


First few reviews:
      filename                                            content
0  review2.txt  I did not enjoy this movie. The storyline was ...
1  review3.txt  An average film with some good moments. Not th...
2  review1.txt  This movie was fantastic! The plot was engagin...


#### Directory of Files Questions

1. How does the loading of these files compare to the loading of structured data with pandas?  
   - Loading text files from a directory requires manual iteration (using `os.listdir`) and file handling (`open`, `.read()`).  
   - With structured data (like CSV, JSON, or Excel), pandas can directly load the entire dataset with a single command (`pd.read_csv`, `pd.read_json`, `pd.read_excel`).  
   - In other words, text files are unstructured and need preprocessing, while pandas automatically parses structured formats into a dataframe.  

2. Which do you prefer and why?  
   - I prefer loading structured data with pandas because it is faster, simpler, and immediately provides a dataframe that can be analyzed with built-in functions.  
   - However, working with unstructured text files gives more flexibility when analyzing raw content (e.g., movie reviews, logs) and is useful for natural language processing tasks.  
